src/dsp/rde.c

```diff
@@ -1223,3 +1223,272 @@ int decode_rde_multipart_receive_resp(
+	pldm_msgbuf_extract_p(buf, operation_locator_length);
+	pldm_msgbuf_extract_p(buf, request_payload_length);
+
+	if (*operation_locator_length > 0) {
+		rc = pldm_msgbuf_extract_array(buf, *operation_locator_length,
+					       operation_locator,
+					       *operation_locator_length);
+		if (rc) {
+			return pldm_msgbuf_discard(buf, rc);
+		}
+	} else {
+		operation_locator = NULL;
+	}
+	if (*request_payload_length > 0) {
+		rc = pldm_msgbuf_extract_array(buf, *request_payload_length,
+					       request_payload,
+					       *request_payload_length);
+		if (rc) {
+			return pldm_msgbuf_discard(buf, rc);
+		}
+	} else {
+		request_payload = NULL;
+	}
+	return pldm_msgbuf_complete(buf);
+}
+
+LIBPLDM_ABI_STABLE
+int encode_rde_operation_init_resp(
+	uint8_t instance_id, uint8_t completion_code, uint8_t operation_status,
+	uint8_t completion_percentage, uint32_t completion_time_seconds,
+	bitfield8_t *operation_execution_flags, uint32_t result_transfer_handle,
+	bitfield8_t *permission_flags, uint32_t response_payload_length,
+	const char *etag, const uint8_t *response_payload, struct pldm_msg *msg)
+{
+	PLDM_MSGBUF_DEFINE_P(buf);
+	int rc;
+
+	if ((msg == NULL) || (operation_execution_flags == NULL) ||
+	    (permission_flags == NULL)) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+	if ((response_payload_length > 0) && (response_payload == NULL)) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+
+	if (completion_code != PLDM_SUCCESS) {
+		return encode_cc_only_resp(instance_id, PLDM_RDE,
+					   PLDM_RDE_OPERATION_INIT,
+					   completion_code, msg);
+	}
+	// Length should include NULL terminator.
+	size_t str_len = strlen(etag) + 1;
+	if (str_len > 255) {
+		return PLDM_ERROR_INVALID_DATA;
+	}
+
+	struct pldm_header_info header;
+	header.msg_type = PLDM_RESPONSE;
+	header.instance = instance_id;
+	header.pldm_type = PLDM_RDE;
+	header.command = PLDM_RDE_OPERATION_INIT;
+	rc = pack_pldm_header(&header, &(msg->hdr));
+	if (rc != PLDM_SUCCESS) {
+		return rc;
+	}
+
+	rc = pldm_msgbuf_init_errno(buf,
+				    PLDM_RDE_OPERATION_INIT_RESP_FIXED_BYTES,
+				    msg->payload,
+				    PLDM_RDE_OPERATION_INIT_RESP_FIXED_BYTES +
+					    PLDM_RDE_VARSTRING_HEADER_SIZE +
+					    str_len + response_payload_length);
+	if (rc) {
+		return rc;
+	}
+
+	pldm_msgbuf_insert(buf, completion_code);
+	pldm_msgbuf_insert(buf, operation_status);
+	pldm_msgbuf_insert(buf, completion_percentage);
+	pldm_msgbuf_insert(buf, completion_time_seconds);
+	pldm_msgbuf_insert(buf, operation_execution_flags->byte);
+	pldm_msgbuf_insert(buf, result_transfer_handle);
+	pldm_msgbuf_insert(buf, permission_flags->byte);
+	pldm_msgbuf_insert(buf, response_payload_length);
+	pldm_msgbuf_insert_uint8(buf, PLDM_RDE_VARSTRING_UTF_8);
+	pldm_msgbuf_insert(buf, (uint8_t)str_len);
+	rc = pldm_msgbuf_insert_array(buf, str_len, (const uint8_t *)etag,
+				      str_len);
+	if (rc != PLDM_SUCCESS) {
+		return rc;
+	}
+
+	if (response_payload_length > 0) {
```

Comment:
The format of this parameter shall be null (consisting of zero bytes) if the
ResponsePayloadLength above is zero;  please make sure handled this also.